### PR TITLE
QField :heart: satellites

### PR DESCRIPTION
**Pull request by nirvn**
Moving our positioning code to C++ made it super easy to add satellite-related information to our handling of _internal_ GNSS positioning.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ef0f1f89500e496d3fcb31116e01d83c8cf7a4c4:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-de6553a2851b1265b70f004731f033b69bc4bdda-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-de6553a2851b1265b70f004731f033b69bc4bdda-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-de6553a2851b1265b70f004731f033b69bc4bdda-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-de6553a2851b1265b70f004731f033b69bc4bdda-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5bb526134c5780633551cb44aef8db5832fc24ee:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-9e4e9a3495f60311caa6fe102d654f92bbede7b3-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-9e4e9a3495f60311caa6fe102d654f92bbede7b3-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-9e4e9a3495f60311caa6fe102d654f92bbede7b3-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-9e4e9a3495f60311caa6fe102d654f92bbede7b3-x86-android.apk)

File: src/core/positioning/internalgnssreceiver.cpp [hunk 1]
```diff
@@ -103,20 +116,81 @@ void InternalGnssReceiver::handlePositionUpdated( const QGeoPositionInfo &positi
...
+    satellitesID << satellite.satelliteIdentifier();
+  }
+  if ( mSatellitesUsed != satellitesUsed
+       || mSatellitesID != satellitesID )
+  {
+    mSatelliteInformationValid = true;
+    mSatellitesUsed = satellitesUsed;
+    mSatellitesID = satellitesID;
+
+    for ( QgsSatelliteInfo &satelliteInfo : mSatellitesInfo )
```
Comment (m-kuhn) [inline src/core/positioning/internalgnssreceiver.cpp:146]:
```suggestion
    for ( QgsSatelliteInfo &satelliteInfo : qgis::as_const( mSatellitesInfo ) )
```

File: src/core/positioning/internalgnssreceiver.cpp [hunk 2]
```diff
@@ -103,20 +116,81 @@ void InternalGnssReceiver::handlePositionUpdated( const QGeoPositionInfo &positi
...
+
 void InternalGnssReceiver::handleError( QGeoPositionInfoSource::Error positioningError )
 {
   qDebug() << positioningError;
   return;
 }
+
+void InternalGnssReceiver::handleSatelliteError( QGeoSatelliteInfoSource::Error satelliteError )
+{
+  qDebug() << satelliteError;
```
Comment (m-kuhn) [inline src/core/positioning/internalgnssreceiver.cpp:194]: Reset values / set valid = false?
Especially for [`QGeoSatelliteInfoSource::ClosedError`](https://doc.qt.io/qt-5/qgeosatelliteinfosource.html#Error-enum)

File: src/core/positioning/internalgnssreceiver.cpp [hunk 3]
```diff
@@ -103,20 +116,81 @@ void InternalGnssReceiver::handlePositionUpdated( const QGeoPositionInfo &positi
...
                                                             verticalSpeed,
                                                             magneticVariation,
                                                             0, mGeoPositionSource->sourceName() );
     emit lastGnssPositionInformationChanged( mLastGnssPositionInformation );
   }
 }
 
+void InternalGnssReceiver::handleSatellitesInUseUpdated( const QList<QGeoSatelliteInfo> &satellites )
+{
+  const int satellitesUsed = satellites.size();
```
Comment (m-kuhn) [inline src/core/positioning/internalgnssreceiver.cpp:133]: According to the docs https://doc.qt.io/qt-5/qgeosatelliteinfosource.html#Error-enum

> If satellite information cannot be retrieved or some other form of timeout has occurred the [satellitesInViewUpdated](https://doc.qt.io/qt-5/qgeosatelliteinfosource.html#satellitesInViewUpdated)() and [satellitesInUseUpdated](https://doc.qt.io/qt-5/qgeosatelliteinfosource.html#satellitesInUseUpdated)() signals may be emitted with empty parameter lists.

we should reset all values if `satellites.isEmpty()`

Comment (nirvn) [reply src/core/positioning/internalgnssreceiver.cpp:133]: Hm, good question. I'm wondering whether we should actually just skip here, to be tolerant (I'm not sure if we need to be here but for positioning we have to as info can go missing for a given reading)

File: src/core/positioning/internalgnssreceiver.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/positioning/internalgnssreceiver.cpp:194]: @m-kuhn , not a bad idea.

File: src/core/positioning/internalgnssreceiver.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/core/positioning/internalgnssreceiver.cpp:146]:
```suggestion
    for ( QgsSatelliteInfo &satelliteInfo : mSatellitesInfo ) // Not const, is modified in loop
```

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 90091a172255014ac2d3174ed32a563da489c716:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-22fdcf3a451c88a997b6c3717f103c0b45e2ccd9-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-22fdcf3a451c88a997b6c3717f103c0b45e2ccd9-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-22fdcf3a451c88a997b6c3717f103c0b45e2ccd9-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3039-22fdcf3a451c88a997b6c3717f103c0b45e2ccd9-x86-android.apk)

**Comment by nirvn** [comment]
@m-kuhn , comments addressed.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**